### PR TITLE
Add InvalidMiddlewareDefinitionException + Improve tests + Update dev dependencies

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

File: src/InvalidMiddlewareDefinitionException.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     * @param array|callable|string $middlewareDefinition
+     */
+    public function __construct($middlewareDefinition)
+    {
+        $message = 'Parameter should be either PSR middleware class name or a callable.';
+
+        $definition = $this->convertDefinitionToString($middlewareDefinition);
+        if ($definition !== null) {
+            $message .= ' Got it: ' . $definition . '.';
+        }
```
Comment (samdark) [inline src/InvalidMiddlewareDefinitionException.php:21]:
```suggestion
        $definitionString = $this->convertDefinitionToString($middlewareDefinition);
        if ($definitionString !== null) {
            $message .= ' Got ' . $definitionString . '.';
        }
```

File: src/InvalidMiddlewareDefinitionException.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        parent::__construct($message);
+    }
+
+    private function convertDefinitionToString($middlewareDefinition): ?string
+    {
+        if (is_object($middlewareDefinition)) {
+            return 'Instance of class "' . get_class($middlewareDefinition) . '"';
```
Comment (samdark) [inline src/InvalidMiddlewareDefinitionException.php:29]:
```suggestion
            return 'an instance of "' . get_class($middlewareDefinition) . '"';
```

File: tests/InvalidMiddlewareDefinitionExceptionTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * @dataProvider dataBase
+     *
+     * @param mixed $definition
+     * @param string $expected
+     */
+    public function testBase($definition, string $expected): void
+    {
+        $exception = new InvalidMiddlewareDefinitionException($definition);
+        $this->assertStringEndsWith('. Got it: ' . $expected . '.', $exception->getMessage());
```
Comment (samdark) [inline tests/InvalidMiddlewareDefinitionExceptionTest.php:44]:
```suggestion
        $this->assertStringEndsWith('. Got ' . $expected . '.', $exception->getMessage());
```

File: tests/InvalidMiddlewareDefinitionExceptionTest.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public function dataBase(): array
+    {
+        return [
+            [
+                'test',
+                '"test"',
+            ],
+            [
+                new TestController(),
+                'Instance of class "Yiisoft\Middleware\Dispatcher\Tests\Support\TestController"',
```
Comment (samdark) [inline tests/InvalidMiddlewareDefinitionExceptionTest.php:22]:
```suggestion
                'an instance of "Yiisoft\Middleware\Dispatcher\Tests\Support\TestController"',
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/InvalidMiddlewareDefinitionException.php (hunk 1 above)
Comment (vjik) [reply src/InvalidMiddlewareDefinitionException.php]: Done.

File: src/InvalidMiddlewareDefinitionException.php (hunk 2 above)
Comment (vjik) [reply src/InvalidMiddlewareDefinitionException.php]: Done.

File: tests/InvalidMiddlewareDefinitionExceptionTest.php (hunk 3 above)
Comment (vjik) [reply tests/InvalidMiddlewareDefinitionExceptionTest.php]: Done.

File: tests/InvalidMiddlewareDefinitionExceptionTest.php (hunk 4 above)
Comment (vjik) [reply tests/InvalidMiddlewareDefinitionExceptionTest.php]: Done.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**